### ISSUE TITLE
Replication lag metric seems to be broken in 1.19.1

### ISSUE THREAD

**Issue by tbondarchuk** [cloudnative-pg/cloudnative-pg#1814]
After updating from 1.19.0 to 1.19.1 I have constant alarms for cnpg_pg_replication_lag metric for all my db clusters. Restarting cluster resets metric back to zero and then it starts to grow again:
<img width="1702" alt="image" src="https://user-images.githubusercontent.com/3264588/227785137-3112669e-577f-4c15-95d7-750a1defa4d8.png">

and unless I'm misreading this - status shows cluster in a perfectly synchronized state:
```sh
kubectl cnpg status account-postgres
Cluster Summary
Name:               account-postgres
Namespace:          qa
System ID:          7175972819327348754
PostgreSQL Image:   ghcr.io/cloudnative-pg/postgresql:14.3
Primary instance:   account-postgres-1
Status:             Cluster in healthy state
Instances:          2
Ready instances:    2
Current Write LSN:  2/1A000000 (Timeline: 9 - WAL File: 000000090000000200000019)

Certificates Status
Certificate Name              Expiration Date                Days Left Until Expiration
----------------              ---------------                --------------------------
account-postgres-ca           2023-06-02 21:10:31 +0000 UTC  68.25
account-postgres-replication  2023-06-02 21:10:31 +0000 UTC  68.25
account-postgres-server       2023-06-02 21:10:31 +0000 UTC  68.25

Continuous Backup status
First Point of Recoverability:  2023-02-24T00:00:09Z
Working WAL archiving:          OK
WALs waiting to be archived:    0
Last Archived WAL:              000000090000000200000019   @   2023-03-26T00:05:11.121184Z
Last Failed WAL:                -

Streaming Replication status
Name                Sent LSN    Write LSN   Flush LSN   Replay LSN  Write Lag  Flush Lag  Replay Lag  State      Sync State  Sync Priority
----                --------    ---------   ---------   ----------  ---------  ---------  ----------  -----      ----------  -------------
account-postgres-2  2/1A000000  2/1A000000  2/1A000000  2/1A000000  00:00:00   00:00:00   00:00:00    streaming  async       0

Unmanaged Replication Slot Status
No unmanaged replication slots found

Instances status
Name                Database Size  Current LSN  Replication role  Status  QoS         Manager Version  Node
----                -------------  -----------  ----------------  ------  ---         ---------------  ----
account-postgres-1  34 MB          2/1A000000   Primary           OK      BestEffort  1.19.1           ip-10-20-26-37.us-west-1.compute.internal
account-postgres-2  33 MB          2/1A000000   Standby (async)   OK      BestEffort  1.19.1           ip-10-20-26-37.us-west-1.compute.internal
```

I'm happy to provide any extra information if needed.

**Comment by mnencia** [comment]
The default lag metric is based on the difference from the last visible commit to the current time. It looks like your cluster is idle, so the metric is unreliable.

**Comment by tbondarchuk** [comment]
@mnencia Well yes, it's dev/qa clusters with no activity when nothing is being tested. Though similar situation can happen in prod as well, if service is using db seldom and not every second, like storing some data which is retrieved once a while.

I guess if is an expected behavior I'll just need to remove that specific alert rule.

But I don't understand why alarms started raising only after 1.19.1 update: was is broken before and fixed in this release or vice versa? Nothing in changelog seems to be related.

**Comment by mnencia** [comment]
That's a good question. The only change I remember in the monitoring code is https://github.com/cloudnative-pg/cloudnative-pg/pull/1666, but I do not see how it could lead to the behavior you see.

**Comment by ondrej-smola** [comment]
Do you use helm chart @tbondarchuk? I seed diff when going to 1.17.0 (1.19.0) to 1.17.1 (1.19.1)

![image](https://user-images.githubusercontent.com/851390/228505135-0ffe6650-1480-438a-b4cf-3733294a9a80.png)

**Comment by osnabrugge** [comment]
I'm seeing the exact same behavior, but to be fair there is not much load on my cluster either.
![image](https://user-images.githubusercontent.com/18647660/229372722-784d75ad-32cb-499f-91ae-021f25a205be.png)
![image](https://user-images.githubusercontent.com/18647660/229372728-c9570655-62c1-471f-8d79-3b43e462fe47.png)

**Comment by tbondarchuk** [comment]
@mnencia I was able to track this to following [chart update](https://github.com/cloudnative-pg/charts/commit/6eeb80eb59dfdec14318e1d20d783a74a04ead9f#diff-a551a220ad6f272352960c48fdc0558ea801fe2f3c551c91e5aa1c97c86963cdR239) - replication lag query[ used to](https://github.com/cloudnative-pg/charts/blob/3a2cdf44f33a3358c1875d1f243cc7859d1fddc4/charts/cloudnative-pg/values.yaml#L240) run on primary. 

Not sure what was the reason to remove primary flag, but judging by [this](https://github.com/prometheus-community/postgres_exporter/issues/385) and [this](https://github.com/prometheus-community/postgres_exporter/pull/534/files) it seems to me that replication lag query has to run on primary? I'm not familiar with postgres replication internals but I believe if alarm is on for idle cluster which looks perfectly replicated otherwise - it must be wrong.

for now I'll simply add `primary: true` back to monitoring configmap but could it be reverted back in chart values? I'd rather avoid adding whole monitoring block to values override in my gitops repo.

**Comment by sxd** [comment]
@mnencia I think this it's fixed right? have you tried the latest dev version @tbondarchuk  ?

**Comment by Zempashi** [comment]
I'm using cloudnative-pg 1.20.2 released 3 weeks ago according to github, and seeing the very same behaviour.

Looking at https://github.com/cloudnative-pg/artifacts/blob/main/manifests/operator-manifest.yaml or the helm chart, I don't see the `primary: true` back, is that the correct fix ?

**Comment by monotek** [comment]
Does "primary: true" mean that the alarm is only used on the primary?
If so, thats imho not the solution as replication lag on the primary should always be 0, or not?

**Comment by B1ue-W01f** [comment]
> Does "primary: true" mean that the alarm is only used on the primary? If so, thats imho not the solution as replication lag on the primary should always be 0, or not?

Doesnt the stat track replication lag to replicas, so yes this is the solution?

Really hoping so because until finding this today ive spent the last few months staring at databases that all appear to be clones of eachother, getting told theyre not replicating, and questioning reality :D

**Comment by B1ue-W01f** [comment]
Yep that was it, deployed and resolved now.

**Comment by B1ue-W01f** [comment]
Created a pull request for this that includes the solution

**Comment by gbartolini** [comment]
The query is supposed to work both on the primary and the standby. We should not use `primary: true`. The SQL statement contains a `CASE` that returns 0 if the instance is not in recovery (ergo, it is a primary, not a standby).

I tried running the query on both a standby and a replica, and it works:

```
$ kubectl exec -ti pg16-1 -c postgres -- psql
psql (16beta3 (Debian 16~beta3-1.pgdg110+2))
Type "help" for help.

postgres=# SELECT CASE WHEN NOT pg_catalog.pg_is_in_recovery()
              THEN 0
              ELSE GREATEST (0,
                EXTRACT(EPOCH FROM (now() - pg_catalog.pg_last_xact_replay_timestamp())))
              END AS lag,
              pg_catalog.pg_is_in_recovery() AS in_recovery,
              EXISTS (TABLE pg_stat_wal_receiver) AS is_wal_receiver_up,
              (SELECT count(*) FROM pg_stat_replication) AS streaming_replicas;
 lag | in_recovery | is_wal_receiver_up | streaming_replicas
-----+-------------+--------------------+--------------------
   0 | f           | f                  |                  2
(1 row)

$ kubectl exec -ti pg16-2 -c postgres -- psql
psql (16beta3 (Debian 16~beta3-1.pgdg110+2))
Type "help" for help.

postgres=# SELECT CASE WHEN NOT pg_catalog.pg_is_in_recovery()
              THEN 0
              ELSE GREATEST (0,
                EXTRACT(EPOCH FROM (now() - pg_catalog.pg_last_xact_replay_timestamp())))
              END AS lag,
              pg_catalog.pg_is_in_recovery() AS in_recovery,
              EXISTS (TABLE pg_stat_wal_receiver) AS is_wal_receiver_up,
              (SELECT count(*) FROM pg_stat_replication) AS streaming_replicas;
     lag      | in_recovery | is_wal_receiver_up | streaming_replicas
--------------+-------------+--------------------+--------------------
 16715.884655 | t           | t                  |                  0
(1 row)
```

In my opinion, we should close this issue, without a patch.

**Comment by Zempashi** [comment]
But then this raise an alarm if we use https://github.com/cloudnative-pg/cloudnative-pg/blob/e133fd93d884f9a1fd2ec048a676988274c8b59a/docs/src/samples/monitoring/prometheusrule.yaml#L41C9-L41C32;

Should we create another issue for addressing that ?

**Comment by gbartolini** [comment]
I am actually about to propose a patch.

**Comment by gbartolini** [comment]
I think the query should look like this:

```sql
SELECT CASE WHEN (
                NOT pg_is_in_recovery()
                OR pg_last_wal_receive_lsn() = pg_last_wal_replay_lsn())
              THEN 0
              ELSE GREATEST (0,
                EXTRACT(EPOCH FROM (now() - pg_last_xact_replay_timestamp())))
              END AS lag,
              pg_is_in_recovery() AS in_recovery,
              EXISTS (TABLE pg_stat_wal_receiver) AS is_wal_receiver_up,
              (SELECT count(*) FROM pg_stat_replication) AS streaming_replicas
```

**Comment by Zempashi** [comment]
Tested on my cluster, it works ;)

**Comment by gbartolini** [comment]
I updated only the documentation stuff. I need to update also the configmap.

**Comment by vitobotta** [comment]
Hey everyone,

I’m facing this issue with a fresh cluster set up using version 1.25.1. Since it's a dev cluster, there isn't much going on in the database—sometimes it’s completely idle. Occasionally, I see that the replication lag on the Grafana dashboard is in the several seconds range. Is this normal?

Thanks!

**Comment by jeremych1000** [comment]
We have an alert configured for this but we do also have several clusters with very little writes. Is there anything we can do on Grafana's side to only alert for the master?
